astropy/coordinates/baseframe.py

```diff
@@ -1750,13 +1757,41 @@ def __ne__(self, value):
         return np.logical_not(self == value)
 
     def _prepare_unit_sphere_coords(
-        self, other: BaseCoordinateFrame | SkyCoord
+        self, other: BaseCoordinateFrame | SkyCoord, frame_mismatch: StrictnessControl
     ) -> tuple[Longitude, Latitude, Longitude, Latitude]:
+        other_frame = getattr(other, "frame", other)
+        if not (
+            frame_mismatch == "ignore"
+            or self.is_equivalent_frame(other_frame)
+            or all(
+                isinstance(comp, (StaticMatrixTransform, DynamicMatrixTransform))
+                for comp in frame_transform_graph.get_transform(
+                    type(self), type(other_frame)
+                ).transforms
+            )
+        ):
+            if frame_mismatch == "warn":
+                warnings.warn(
+                    AstropyUserWarning(
+                        "transforming other coordinates from "
```

Comment:
I think the message needs to indicate why this is a problem - for a user, it may well happen fairly deep inside a library. Something like
```
calculating the separation requires transforming between
 ...self... and ...other...,
which is not a simple rotation. Hence the result may depend on the frame one
starts from.
``` 

It may also be good to define the main message up front (as a regular format string, not an f-string), so it can just be filled in here and for the error.

Reply:
I've added a new error and a new warning with custom `__str__()` methods, which takes care of specifying the error message.